src/bounded/sat.rs

```diff
@@ -0,0 +1,811 @@
+            _ => return Err(CheckerError::OutOfOrderStatement(statement.clone())),
+        }
+    }
+
+    let mut inits = Vec::new();
+    let mut trs = Vec::new();
+    for assume in assumes {
+        match assume {
+            Term::UnaryOp(UOp::Always, axiom) if FirstOrder::unrolling(&axiom) == Some(0) => {
+                inits.push(*axiom.clone());
+                trs.push(*axiom);
+            }
+            Term::UnaryOp(UOp::Always, tr) if FirstOrder::unrolling(&tr) == Some(1) => {
+                trs.push(*tr)
+            }
+            Term::UnaryOp(UOp::Always, term) => return Err(CheckerError::TooFuture(*term)),
+            init if FirstOrder::unrolling(&init) == Some(0) => inits.push(init),
+            init => return Err(CheckerError::AnyFuture(init)),
+        }
+    }
+
+    let mut safes = Vec::new();
+    for assert in asserts {
+        match assert {
+            Term::UnaryOp(UOp::Always, safe) if FirstOrder::unrolling(&safe) == Some(0) => {
+                safes.push(*safe)
+            }
+            Term::UnaryOp(UOp::Always, safe) => return Err(CheckerError::AnyFuture(*safe)),
+            assert => return Err(CheckerError::AssertWithoutAlways(assert)),
+        }
+    }
+
+    let mut context = Context::new(&module.signature, universe, depth);
+
+    let translate = |terms| {
+        let term = Term::NAryOp(NOp::And, terms);
+        let term = nullary_id_to_app(term, &module.signature.relations);
+        let term = crate::term::Next::new(&module.signature).normalize(&term);
+        term_to_ast(&term, &context, &HashMap::new())
+    };
+
+    println!("starting translation...");
+    let translation = std::time::Instant::now();
+
+    let init = translate(inits)?;
+    let tr = translate(trs)?;
+    let not_safe = Ast::Not(Box::new(translate(safes)?));
+
+    let mut program = vec![init];
+    for i in 0..depth {
+        program.push(tr.clone().prime(i));
+    }
+    program.push(not_safe.prime(depth));
+
+    let cnf = tseytin(&Ast::And(program), &mut context);
+
+    println!("translation finished in {:?}", translation.elapsed());
+
+    println!("starting cadical search...");
+    let cadical_search = std::time::Instant::now();
+
+    let mut solver: cadical::Solver = Default::default();
+    for clause in &cnf {
+        let cadical_clause = clause
+            .iter()
+            .map(|l| (l.var as i32 + 1) * if l.pos { 1 } else { -1 });
+        solver.add_clause(cadical_clause);
+    }
+    assert_eq!(solver.max_variable(), context.vars as i32);
+
+    let cadical_answer = match solver.solve() {
+        None => return Err(CheckerError::SolverFailed),
+        Some(false) => CheckerAnswer::Unknown,
+        Some(true) => {
+            context.print_counterexample(solver, depth);
+            CheckerAnswer::Counterexample
+        }
+    };
+
+    println!("cadical search finished in {:?}", cadical_search.elapsed());
```

Comment:
I would prefer this print in seconds with `{:0.1f}s` to avoid excessive precision.